lightning/src/offers/invoice.rs

```diff
@@ -0,0 +1,633 @@
+	fee_proportional_millionths: u32,
+	cltv_expiry_delta: u16,
+	htlc_minimum_msat: u64,
+	htlc_maximum_msat: u64,
+	features: BlindedHopFeatures,
+}
+
+impl_writeable!(BlindedPayInfo, {
+	fee_base_msat,
+	fee_proportional_millionths,
+	cltv_expiry_delta,
+	htlc_minimum_msat,
+	htlc_maximum_msat,
+	features
+});
+
+/// Wire representation for an on-chain fallback address.
+#[derive(Debug, PartialEq)]
+pub(super) struct FallbackAddress {
+	version: WitnessVersion,
+	program: Vec<u8>,
+}
+
+impl_writeable!(FallbackAddress, { version, program });
+
+type FullInvoiceTlvStream =
+	(PayerTlvStream, OfferTlvStream, InvoiceRequestTlvStream, InvoiceTlvStream, SignatureTlvStream);
+
+impl SeekReadable for FullInvoiceTlvStream {
+	fn read<R: io::Read + io::Seek>(r: &mut R) -> Result<Self, DecodeError> {
+		let payer = SeekReadable::read(r)?;
+		let offer = SeekReadable::read(r)?;
+		let invoice_request = SeekReadable::read(r)?;
+		let invoice = SeekReadable::read(r)?;
+		let signature = SeekReadable::read(r)?;
+
+		Ok((payer, offer, invoice_request, invoice, signature))
+	}
+}
+
+type PartialInvoiceTlvStream =
+	(PayerTlvStream, OfferTlvStream, InvoiceRequestTlvStream, InvoiceTlvStream);
+
+type PartialInvoiceTlvStreamRef<'a> = (
+	PayerTlvStreamRef<'a>,
+	OfferTlvStreamRef<'a>,
+	InvoiceRequestTlvStreamRef<'a>,
+	InvoiceTlvStreamRef<'a>,
+);
+
+impl TryFrom<ParsedMessage<FullInvoiceTlvStream>> for Invoice {
+	type Error = ParseError;
+
+	fn try_from(invoice: ParsedMessage<FullInvoiceTlvStream>) -> Result<Self, Self::Error> {
+		let ParsedMessage { bytes, tlv_stream } = invoice;
+		let (
+			payer_tlv_stream, offer_tlv_stream, invoice_request_tlv_stream, invoice_tlv_stream,
+			SignatureTlvStream { signature },
+		) = tlv_stream;
+		let contents = InvoiceContents::try_from(
+			(payer_tlv_stream, offer_tlv_stream, invoice_request_tlv_stream, invoice_tlv_stream)
+		)?;
+
+		let signature = match signature {
+			None => return Err(ParseError::InvalidSemantics(SemanticError::MissingSignature)),
+			Some(signature) => signature,
+		};
+		let pubkey = contents.fields().signing_pubkey;
+		merkle::verify_signature(&signature, SIGNATURE_TAG, &bytes, pubkey)?;
+
+		Ok(Invoice { bytes, contents, signature })
+	}
+}
+
+impl TryFrom<PartialInvoiceTlvStream> for InvoiceContents {
+	type Error = SemanticError;
+
+	fn try_from(tlv_stream: PartialInvoiceTlvStream) -> Result<Self, Self::Error> {
+		let (
+			payer_tlv_stream,
+			offer_tlv_stream,
+			invoice_request_tlv_stream,
+			InvoiceTlvStream {
+				paths, blindedpay, created_at, relative_expiry, payment_hash, amount, fallbacks,
+				features, node_id,
+			},
+		) = tlv_stream;
+
+		let (paths, payinfo) = match (paths, blindedpay) {
+			(None, _) => return Err(SemanticError::MissingPaths),
+			(_, None) => return Err(SemanticError::InvalidPayInfo),
+			(Some(paths), _) if paths.is_empty() => return Err(SemanticError::MissingPaths),
+			(Some(paths), Some(blindedpay)) if paths.len() != blindedpay.len() => {
```

Comment:
We can still have a single parameter in `InvoiceBuilder::new` with something like type `Vec<(BlindedPath, BlindedPayInfo)>`. But I think the question then is if there should be a single accessor if it is represented internally using two `Vec`s. The return type would be fed into the path finding algorithm. That accessor could return `Vec<(&BlindedPath, &BlindedPayInfo)>`, which -- while requiring an allocation -- wouldn't need any clones and would be better than what we do for BOLT 11 hints and `ChannelDetails`.

Reply:
> Hmm, right, annoying we'll have the same problem when we go to pass through to route finding.
> 
> I wonder if there's a way we can implement the writing on an iterator instead of on vecs themselves to avoid this whole mess?

Took a stab at this but ran into some lifetime problems. See last commit. Not sure if it's the best approach or worth pursuing at the moment. Can be revisit later, IMO.

Reply:
Ah, right. Reading needs to use `WithoutLength` still and store a `Vec` in `InvoiceTlvStream`. I managed to get it working with some changes in the decoding side of the macros. PTAL.